src/compiler/checker.ts

```diff
@@ -28873,7 +28859,7 @@ namespace ts {
             if (returnType.flags & TypeFlags.ESSymbolLike && isSymbolOrSymbolForCall(node)) {
                 return getESSymbolLikeTypeForNode(walkUpParenthesizedExpressions(node.parent));
             }
-            if (node.kind === SyntaxKind.CallExpression && node.parent.kind === SyntaxKind.ExpressionStatement &&
+            if (node.kind === SyntaxKind.CallExpression && !node.questionDotToken && node.parent.kind === SyntaxKind.ExpressionStatement &&

```

Comment:
Was this supposed to be part of this change set? Or part of another PR?

Reply:
It's an intended change. We have some code in `checkCallExpression` that attempts to issue errors when calls to assertion functions aren't actually processed as assertions (because, for example, they don't have targets that were declared with explicit type annotations). This code kicks in only with a function's return type is exactly `void`. It would not kick in in cases where the return type was `void | undefined`, but now that we reduce that to `void`, it does kick in.

The core issue really is that we don't want the check when the call expression is a question-dot call because the assertion isn't know to have actually happened. So that's what this fixes.